### PR TITLE
Fluidigm Analysis File Creator

### PR DESCRIPTION
**Pull request by jbeast**
EPP to create a file that is needed to load into the Fluidigm software
for analysis

### CODE REVIEW

File: src/perl/lib/wtsi_clarity/epp/sm/fluidigm_analysis_file_creator.pm [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+                          ->input_artifacts
+                          ->findnodes(sprintf $ARTIFACT_BY_SAMPLE_VALUE, $clarity_well);
+
+      $sample{'well_location'} = $fluidigm_well;
+
+      if ($artifact_list->size == 0) {
+        $sample{'sample_name'} = '[ Empty ]';
+
+        if ($sample{'well_location'} eq 'H12') {
+          $sample{'sample_type'} = 'NTC';
```
Comment (ke4) [inline src/perl/lib/wtsi_clarity/epp/sm/fluidigm_analysis_file_creator.pm]: I would create a Readonly constant for 'NTC' with a meaningful name (right now I don't know what NTC is ;)).

File: src/perl/lib/wtsi_clarity/epp/sm/fluidigm_analysis_file_creator.pm [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+        }
+      } elsif ($artifact_list->size == 1) {
+        $sample{'sample_name'} = $artifact_list->pop->findvalue($SAMPLE_LIMSID);
+      }
+
```
Comment (ke4) [inline src/perl/lib/wtsi_clarity/epp/sm/fluidigm_analysis_file_creator.pm]: What happens if artifact_list contains more than 1 element?

File: src/perl/lib/wtsi_clarity/epp/sm/fluidigm_analysis_file_creator.pm (hunk 1 above)
Comment (jbeast) [inline src/perl/lib/wtsi_clarity/epp/sm/fluidigm_analysis_file_creator.pm]: Okay will do (although the full version doesn't make much sense either... "No Template Control")

Comment (ke4) [inline src/perl/lib/wtsi_clarity/epp/sm/fluidigm_analysis_file_creator.pm]: I just meant: a new dev won't know what it is and why the code is doing this. Maybe you could add a comment?

File: src/perl/lib/wtsi_clarity/util/csv/factory.pm [hunk 3]
```diff
@@ -27,6 +27,8 @@ sub create {
     $csv_factory = wtsi_clarity::util::csv::factories::generic_csv_reader->new();
   } elsif ( $csv_type eq 'calliper_reader') {
     $csv_factory = wtsi_clarity::util::csv::factories::calliper_csv_reader->new();
+  } elsif ( $csv_type eq 'fluidigm_writer') {
+    $csv_factory = wtsi_clarity::util::csv::factories::generic_csv_writer->new();
```
Comment (ke4) [inline src/perl/lib/wtsi_clarity/util/csv/factory.pm]: It is getting very ugly.
If we would have time I would refactor it using the Strategy Pattern: https://en.wikipedia.org/wiki/Strategy_pattern

File: src/perl/lib/wtsi_clarity/epp/sm/fluidigm_analysis_file_creator.pm (hunk 2 above)
Comment (jbeast) [inline src/perl/lib/wtsi_clarity/epp/sm/fluidigm_analysis_file_creator.pm]: I've made the assumption that Clarity would never give us a plate with duplicate wells on it, but I'll add croak anyway

**Pull request merged**